internal/scheduler/levenshtein_test.go

```diff
@@ -0,0 +1,196 @@
+			b:        []string{"host1", "host2"},
+			expected: 1,
+			edits: []LevEdit{
+				LevDelete("host3", 2),
+			},
+		},
+		{
+			name:     "One substitution",
+			a:        []string{"host1", "host2", "host3"},
+			b:        []string{"host1", "hostX", "host3"},
+			expected: 1,
+			edits: []LevEdit{
+				LevSub("hostX", 1),
+			},
+		},
+		{
+			name:     "Empty slices",
+			a:        []string{},
+			b:        []string{},
+			expected: 0,
+			edits:    []LevEdit{},
+		},
+		{
+			name:     "One empty slice",
+			a:        []string{"host1", "host2", "host3"},
+			b:        []string{},
+			expected: 3,
+			edits: []LevEdit{
+				LevDelete("host1", 0),
+				LevDelete("host2", 1),
+				LevDelete("host3", 2),
+			},
+		},
+		{
+			name:     "Reordering",
+			a:        []string{"host1", "host2", "host3"},
+			b:        []string{"host3", "host2", "host1"},
+			expected: 2,
+			edits: []LevEdit{
+				LevSub("host3", 0),
+				LevSub("host1", 2),
+			},
+		},
+		{
+			name:     "Reordering cascaded",
+			a:        []string{"host1", "host2", "host3", "host4", "host5"},
+			b:        []string{"host4", "host1", "host2", "host3", "host5"},
+			expected: 2, // Only host 4 is moved to the front
+			edits: []LevEdit{
+				LevInsert("host4", 0),
+				LevDelete("host4", 3),
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			result, edits := levenshtein(tt.a, tt.b)
+			if result != tt.expected {
+				t.Errorf("levenshtein(%v, %v) = %d; want %d", tt.a, tt.b, result, tt.expected)
+			}
+			if len(edits) != len(tt.edits) {
+				t.Errorf("levenshtein(%v, %v) edits = %v; want %v", tt.a, tt.b, edits, tt.edits)
+			}
+			for i, edit := range edits {
+				if edit != tt.edits[i] {
+					t.Errorf("levenshtein(%v, %v) edit %d = %v; want %v", tt.a, tt.b, i, edit, tt.edits[i])
+				}
+			}
+		})
+	}
+}
+
+func TestHostShift(t *testing.T) {
+	tests := []struct {
+		name     string
+		a        []string
+		edits    []LevEdit
+		expected int
+	}{
+		{
+			name:     "No edits",
+			a:        []string{"host1", "host2", "host3"},
+			edits:    []LevEdit{},
+			expected: 0,
+		},
+		{
+			name: "Single insertion",
+			a:    []string{"host1", "host2"},
+			edits: []LevEdit{
+				LevInsert("host3", 2),
+			},
+			expected: 0, // Insertions alone are not counted
+		},
+		{
+			name: "Single deletion",
+			a:    []string{"host1", "host2", "host3"},
+			edits: []LevEdit{
+				LevDelete("host3", 2),
+			},
+			expected: 0, // Deletions alone are not counted
+		},
+		{
+			name: "Insertion and deletion of the same host",
+			a:    []string{"host1", "host2", "host3"},
+			edits: []LevEdit{
+				LevInsert("host3", 0),
+				LevDelete("host3", 2),
+			},
+			expected: 2, // Host shifted by 2 positions
+		},
+		{
+			name: "Substitution pair",
+			a:    []string{"host1", "host2", "host3"},
+			edits: []LevEdit{
+				LevSub("host3", 0),
+				LevSub("host1", 2),
+			},
+			expected: 4, // Swapped distance counted twice
```

Comment:
My fluency in the Levenshtein algorithm is pretty rusty.
Is the transformation [1,2,3] -> [3,2,1] determined to be swap only rather than 2 shifts [1,2,3] -> [3,1,2] -> [3,2,1] because that transform nominally consists of fewer transformations? 2 substitutions vs. 2 deletions + 2 insertions.

Asking because - if I understand the hostShift algo correctly - the 2 swaps would have a shift value of 4 while the 2 deletions + 2 insertions would have a shift value of 3.

Reply:
But I definitely need to check if levenshtein calculates the insertions + deletions on the static string, meaning that the value would be 4 contrasting your example where you assume transitive edits on the string: 
[a, b, c] to [c, b, a] -> del a[0] ins a[2], del c[2] ins c[0] -> shift 4

Reply:
After reconsidering, there may be more cases where the shift metric would be broken, e.g.
```
levenshtein(
  [host1 host2 host3 host4 host5 host6], 
  [host4 host1 host2 host3 host6 host5]
) 
edits = [{insert host4 0} {sub host6 4} {delete host6 5}]
```
So I need to come up with something different.